core/src/broadcast_stage/broadcast_duplicates_run.rs

```diff
@@ -0,0 +1,333 @@
+        if last_tick_height == bank.max_tick_height() {
+            self.next_shred_index = 0;
+            self.duplicate_queue
+                .register_hash(&self.last_duplicate_entry_hash, &FeeCalculator::default());
+        } else {
+            self.next_shred_index = last_shred_index;
+        }
+
+        // Partition network with duplicate and real shreds based on stake
+        let bank_epoch = bank.get_leader_schedule_epoch(bank.slot());
+        let mut duplicate_recipients = HashMap::new();
+        let mut real_recipients = HashMap::new();
+
+        let mut stakes: Vec<(Pubkey, u64)> = bank
+            .epoch_staked_nodes(bank_epoch)
+            .unwrap()
+            .into_iter()
+            .filter(|(pubkey, _)| *pubkey != self.keypair.pubkey())
+            .collect();
+        stakes.sort_by(|(l_key, l_stake), (r_key, r_stake)| {
+            if r_stake == l_stake {
+                l_key.cmp(&r_key)
+            } else {
+                r_stake.cmp(&l_stake)
+            }
+        });
+
+        let highest_staked_node = stakes.first().cloned().map(|x| x.0);
+        let stake_total: u64 = stakes.iter().map(|(_, stake)| *stake).sum();
+        let mut cumulative_stake: u64 = 0;
+        for (pubkey, stake) in stakes.into_iter().rev() {
+            cumulative_stake += stake;
+            if (100 * cumulative_stake / stake_total) as u8 <= self.config.stake_partition {
+                duplicate_recipients.insert(pubkey, stake);
+            } else {
+                real_recipients.insert(pubkey, stake);
+            }
+        }
+
+        if let Some(highest_staked_node) = highest_staked_node {
+            if bank.slot() > MINIMUM_DUPLICATE_SLOT && last_tick_height == bank.max_tick_height() {
+                warn!(
+                    "{} sent duplicate slot {} to nodes: {:?}",
+                    self.keypair.pubkey(),
+                    bank.slot(),
+                    &duplicate_recipients,
+                );
+                warn!(
+                    "Duplicate shreds for slot {} will be broadcast in {} slot(s)",
+                    bank.slot(),
+                    self.config.duplicate_send_delay
+                );
+
+                let delayed_shreds: Option<Vec<Shred>> = vec![
+                    duplicate_data_shreds.last().cloned(),
+                    data_shreds.last().cloned(),
+                ]
+                .into_iter()
+                .collect();
+                self.delayed_queue
+                    .lock()
+                    .unwrap()
+                    .push_back((Some(highest_staked_node), delayed_shreds));
+            }
+        }
+
+        let duplicate_recipients = Arc::new(duplicate_recipients);
+        let real_recipients = Arc::new(real_recipients);
+
+        let data_shreds = Arc::new(data_shreds);
+        blockstore_sender.send((data_shreds.clone(), None))?;
+
+        // 3) Start broadcast step
```

Comment:
For my understanding, why do we send out the duplicate shreds first before the real recipients?